kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/EngineUIProxyServlet.scala

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.server.api
+
+import javax.servlet.http.HttpServletRequest
+
+import org.eclipse.jetty.client.api.Request
+import org.eclipse.jetty.proxy.ProxyServlet
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.server.api.EngineUIProxyServlet.{CONTEXT_HEADER_KEY, ENGINE_UI_PROXY_PATH}
+
+private[api] class EngineUIProxyServlet extends ProxyServlet with Logging {
+
+  override def rewriteTarget(request: HttpServletRequest): String = {
+    var targetUrl = "/no-ui-error"
+    val requestUrl = request.getRequestURI
+    getTargetAddress(requestUrl).foreach {
+      case (host, port) =>
+        val subPath = s"/$ENGINE_UI_PROXY_PATH/$host:$port/"
+        val targetPath = requestUrl.substring(subPath.length)
+        targetUrl = s"http://$host:$port/$targetPath"
+    }
+    debug(s"rewrite $requestUrl => $targetUrl")
+    targetUrl
+  }
+
+  override def addXForwardedHeaders(
+      clientRequest: HttpServletRequest,
+      proxyRequest: Request): Unit = {
+    val addressPair = getTargetAddress(clientRequest.getRequestURI)
+    addressPair.foreach { case (host, port) =>
+      proxyRequest.header(CONTEXT_HEADER_KEY, s"/engine-ui/$host:$port")
```

Comment:
why this is required? it does not like a standard http header

Reply:
We can ref to Spark Doc about  `spark.ui.proxyRedirectUri`

> Where to address redirects when Spark is running behind a proxy. This will make Spark modify redirect responses so they point to the proxy server, instead of the Spark UI's own address. This should be only the address of the server, without any prefix paths for the application; the prefix should be set either by the proxy server itself (by adding the X-Forwarded-Context request header), or by setting the proxy base in the Spark app's configuration.

Reply:
OK, let's add the link in comment